kaspad/src/daemon.rs

```diff
@@ -308,13 +312,106 @@ do you confirm? (answer y/n or pass --yes to the Kaspad command line to confirm
+
+        // TODO: Update this entire section to a more robust implementation that allows applying multiple upgrade strategies.
+        // If I'm at version 3 and latest version is 7, I need to be able to upgrade to that version following the intermediate
+        // steps without having to delete the DB
+        if version == 3 {
+            let active_consensus_dir_name = mcms.active_consensus_dir_name().unwrap();
+
+            match active_consensus_dir_name {
+                Some(current_consensus_db) => {
+                    // Apply soft upgrade logic: delete GD data from higher levels
+                    // and then update DB version to 4
+                    let consensus_db = kaspa_database::prelude::ConnBuilder::default()
+                        .with_db_path(consensus_db_dir.clone().join(current_consensus_db))
+                        .with_files_limit(1)
+                        .build()
+                        .unwrap();
+                    info!("Scanning for deprecated records to cleanup");
+
+                    let mut gd_record_count: u32 = 0;
+                    let mut compact_record_count: u32 = 0;
+
+                    let start_level: u8 = 1;
+                    let start_level_bytes = start_level.to_le_bytes();
+                    let ghostdag_prefix_vec = DatabaseStorePrefixes::Ghostdag.into_iter().chain(start_level_bytes).collect_vec();
+                    let ghostdag_prefix = ghostdag_prefix_vec.as_slice();
+
+                    // This section is used to count the records to be deleted. It's not used for the actual delete.
+                    for result in consensus_db.iterator(rocksdb::IteratorMode::From(ghostdag_prefix, rocksdb::Direction::Forward)) {
+                        let (key, _) = result.unwrap();
+                        if !key.starts_with(&[DatabaseStorePrefixes::Ghostdag.into()]) {
+                            break;
+                        }
+
+                        gd_record_count += 1;
+                    }
+
+                    let compact_prefix_vec = DatabaseStorePrefixes::GhostdagCompact.into_iter().chain(start_level_bytes).collect_vec();
+                    let compact_prefix = compact_prefix_vec.as_slice();
+
+                    for result in consensus_db.iterator(rocksdb::IteratorMode::From(compact_prefix, rocksdb::Direction::Forward)) {
+                        let (key, _) = result.unwrap();
+                        if !key.starts_with(&[DatabaseStorePrefixes::GhostdagCompact.into()]) {
+                            break;
+                        }
+
+                        compact_record_count += 1;
+                    }
+
+                    trace!("Number of Ghostdag records to cleanup: {}", gd_record_count);
+                    trace!("Number of GhostdagCompact records to cleanup: {}", compact_record_count);
+                    info!("Number of deprecated records to cleanup: {}", gd_record_count + compact_record_count);
+
+                    let msg =
```

Comment:
Why don't we ask before going over the db?

Reply:
The code above this question is only there to show to the user that there are indeed some number of records to cleanup. We don't really need to go over the DB since the efficient way to delete it is done below (with ranges). We do so only because we want to show some stats to the user to inform their decision.

We can:
- move this question up, before the stats are shown. The UX is then that we ask the question first, then show the stats to the user, then (without asking again) proceed with the efficient delete
- remove the stats section altogether - the user is not informed of the number of records to be deleted, maybe it doesn't matter to them anyway and they would always have to upgrade if they want to use this version of the node and DB
- keep it as it is

Reply:
I'm ok with it as is. It gives a solid picture to the user and it isn't really expensive